### PR TITLE
[HLSL] Align language modes on 202x as default

### PR DESCRIPTION
**Pull request by llvm-beanz**
As captured in issue #108044, HLSL 202x is the target language mode for conformance for Clang. Earlier language modes will be a best effort and prioritized after 2020x. To make this easier and reduce our testing complexity we want to make 202x the default language mode now, and align all earlier modes to match 202x (except where we explicitly deviate).

This change has the following concrete changes:
* All older language modes gain `CPlusPlus11` as a base
* The default language mode for HLSL sources is changed to 202x
* A few test cases are updated to resolve differences in generated diagnostics.

Second to last change for  #108044

### CODE REVIEW

**Review by farzonl — APPROVED** [review]

File: clang/test/Preprocessor/predefined-macros-hlsl.hlsl [hunk 1]
```diff
@@ -12,7 +12,7 @@
 // HALF: #define __HLSL_ENABLE_16_BIT 1
 // NOHALF-NOT: __HLSL_ENABLE_16_BIT
 
-// CHECK: #define __HLSL_VERSION 2021
+// CHECK: #define __HLSL_VERSION 2028
```
Comment (pow2clk) [inline clang/test/Preprocessor/predefined-macros-hlsl.hlsl:15]: 2028 == 202x and 2029 == 202y I expect?

Comment (bob80905) [inline clang/test/Preprocessor/predefined-macros-hlsl.hlsl:15]: I do not understand where this 2028 is coming from. Is 202x intended to be 2028?

File: clang/test/SemaHLSL/Types/Arithmetic/literal_suffixes_202x.hlsl [hunk 2]
```diff
@@ -37,7 +37,7 @@ _Static_assert(is_same<__decltype(4294967296), int64_t>::value, "4294967296 is i
 // Clang emits a warning that it is interpreting it as unsigned because that is
 // not conforming to the C standard.
 
-// expected-warning@+1{{integer literal is too large to be represented in type 'long' and is subject to undefined behavior under C++98, interpreting as 'unsigned long'; this literal will be ill-formed in C++11 onwards}}
+// expected-warning@+1{{integer literal is too large to be represented in a signed integer type, interpreting as unsigned}}
```
Comment (pow2clk) [inline clang/test/SemaHLSL/Types/Arithmetic/literal_suffixes_202x.hlsl:40]: It's all good now, but even though the filename suggests it's relevant to 202x, it wasn't targeting that before. Perhaps that contributed to the need for this change

**Review by pow2clk — APPROVED** [review]
I'm not objecting to anything, but I made a few comments out of interest

File: clang/test/Preprocessor/predefined-macros-hlsl.hlsl (hunk 1 above)
Comment (llvm-beanz) [reply clang/test/Preprocessor/predefined-macros-hlsl.hlsl:15]: 202x doesn't have a year yet, so to make it big and far out I gave it 2028 and 202y has 2029... these will change before we actually ship.

Comment (llvm-beanz) [reply clang/test/Preprocessor/predefined-macros-hlsl.hlsl:15]: Yea. Subject to revision once we have an actual year.

File: clang/test/SemaHLSL/Types/Arithmetic/literal_suffixes_202x.hlsl (hunk 2 above)
Comment (llvm-beanz) [reply clang/test/SemaHLSL/Types/Arithmetic/literal_suffixes_202x.hlsl:40]: The test is verifying that Clang implements the 202x conforming literals feature. Clang doesn't implement the legacy behavior, so it is really the only behavior in Clang.

**Pull request merged**